### PR TITLE
c-ares support for curl. (py)curl updates.

### PR DESCRIPTION
**Pull request by giffels**
Hi Diego,

as discussed via email, the DBS 3 Migration Service has some problems with pycurl and multi-threading. pycurl itself is thread-safe, however the DNS lookup is not. The suggestion on http://linux.die.net/man/3/libcurl-tutorial is to disable signals, which I did for DBS 3. The disadvantage is, that DNS timeouts are not caught anymore in that case. The solution is to use c-ares for the DNS lookup. I have added a c-ares.spec and configured curl to build against c-ares.

In addition, I have upgraded curl and pycurl to the latest release to enable shared SSL session id caches between different threads, to reduce the load on cmsweb through authentication during migration requests. 

I have tested DBS 3 Server, DBS 3 Migration Service and DBS 3 Client, everything seems to be fine.

Could you review and merge, please?

Thanks,
Manuel

### CODE REVIEW

**Comment by giffels** [comment]
As agreed with Alan we are going to postpone this to the March release, since it is not critical for the operation of DBS and could have impacts to the cmsweb upgrade and other services. 

Cheers,
Manuel

**Comment by amaltaro** [comment]
Hi Manuel, @yuyiguo, I lost track of this pull request. Is DBS3 ready for these changes?

**Comment by giffels** [comment]
Hi Alan,

we would like to include this in the March deployment. Current DBS 3 will work with it, however does not yet profit from it. We need to merge https://github.com/dmwm/DBS/pull/313 before to profit.

Cheers,
Manuel

**Comment by amaltaro** [comment]
@yuyiguo, @giffels , it seems https://github.com/dmwm/DBS/pull/313 was still not accepted and so it is not available in DBS 3.2.2a (version requested to go into this cycle). Are we going to skip it for April's deployment?

**Comment by yuyiguo** [comment]
Hi, Alan:

I did not put it in because I am not sure if you are going to give green light for this or not. We can add it in for next cmsweb-testbed deployment for HG1403.
Cheers,
Yuyi

**Comment by amaltaro** [comment]
Hi Yuyi, if you have time to accept that c-ares change in DBS, test it and update the PR (new DBS tag(?)), then I can get all changes tomorrow early in the morning.

File: c-ares.spec [hunk 1]
```diff
@@ -0,0 +1,23 @@
+### RPM external c-ares 1.10.0
+Source: http://c-ares.haxx.se/download/%n-%realversion.tar.gz
+Provides: libcares.2.so()(64bit)
```
Comment (amaltaro) [inline c-ares.spec:3]: I'll update it because in the end we get 'libcares.so.2' inside the RPM instead of 'libcares.2.so'. Thanks

**Comment by yuyiguo** [comment]
@giffels 
Hi, Alan:

I built a new rpm, but got error when testing.
 DBSMigrateModle/submit. 'module' object has no attribute 'LOCK_DATA_SSL_SESSION'
File "/data/srv/HG1401j/sw.pre.yuyi/slc5_amd64_gcc461/cms/dbs3/3.2.2b/lib/python2.6/site-packages/dbs/business/DBSMigrate.py", line 203, in prepareBlockMigrationList
    raise ex
AttributeError: 'module' object has no attribute 'LOCK_DATA_SSL_SESSION'

I am not sure what correct version of pycurl should be. I am using HG1401j. 
Manuel is off on vacation right now. Please go head using DBS_3_2_2a for tomorrow's deployment.

**Comment by amaltaro** [comment]
Thanks for checking Yuyi. Then I'm proceeding with 3.2.2a without the c-ares library. We can debug and integrate these changes afterwards. Thanks

**Comment by giffels** [comment]
Hi Yuyi,

you have to use the modified spec files from this pull request to build DBS with a new pycurl, curl and c-ares version.

Cheers,
Manuel

**Comment by yuyiguo** [comment]
Thanks Manuel for pointing it out. I will try again with this new spec.

**Comment by yuyiguo** [comment]
@giffels 
Hi, Alan:

I rebuilt and tested with the new/updated specs in this pull request. Now all DBS3 tests are past. I am going to merge the change into DBS3 main branch. Would you please merge these specs to comp too?

Thanks,
Yuyi

**Comment by amaltaro** [comment]
Yes Yuyi! We'll merge and include it for the next testbed HG tag, which is going to happen next Monday/Tuesday. Can you make a new DBS PR by Monday morning please? Thanks

**Comment by yuyiguo** [comment]
OK, Alan, I will make a new tag. What is "PR"?
Yuyi

**Comment by amaltaro** [comment]
Thanks Yuyi!
PR = Pull request :-)

**Comment by yuyiguo** [comment]
Request made : https://svnweb.cern.ch/trac/CMSDMWM/ticket/4282#comment:15

**Comment by amaltaro** [comment]
Done! It went along the HG1403b tag. Thanks

**Pull request closed without merging**